contract_comm/currency/currency.go

```diff
@@ -94,97 +94,127 @@ const (
+	if er == nil && anotherTokenRate == nil {
+		return amount.Cmp(anotherTokenAmount)
 	}
 
-	// Given value of val and rates n1/d1 and n2/d2 the function below does
-	// (val * d1 * n2) / (n1 * d2)
-	numerator := new(big.Int).Mul(val, new(big.Int).Mul(exchangeRateFrom.Denominator, exchangeRateTo.Numerator))
-	denominator := new(big.Int).Mul(exchangeRateFrom.Numerator, exchangeRateTo.Denominator)
+	// if both exchangeRate are the same
+	if er != nil && anotherTokenRate != nil && *er == *anotherTokenRate {
+		return amount.Cmp(anotherTokenAmount)
+	}
 
-	if denominator.Sign() == 0 {
-		log.Error("Convert - Error in retreiving currency exchange rates")
-		return nil, errors.ErrExchangeRateZero
+	// Below code block is basically evaluating this comparison:
+	// amount * er.denominator / er.numerator < anotherTokenAmount * anotherTokenRate.denominator / anotherTokenRate.numerator
+	// It will transform that comparison to this, to remove having to deal with fractional values.
+	// amount * er.denominator * anotherTokenRate.numerator < anotherTokenAmount * anotherTokenRate.denominator * er.numerator
+
+	// when either er or anotherTokenRate are nil, we assume rate = 1/1 (as in cGLD)
+	var leftSide, rightSide *big.Int
+	if er == nil {
+		leftSide = new(big.Int).Mul(amount, anotherTokenRate.numerator)
+		rightSide = new(big.Int).Mul(anotherTokenAmount, anotherTokenRate.denominator)
+	} else if anotherTokenRate == nil {
+		leftSide = new(big.Int).Mul(amount, er.denominator)
+		rightSide = new(big.Int).Mul(anotherTokenAmount, er.numerator)
+	} else {
+		leftSide = new(big.Int).Mul(amount, new(big.Int).Mul(er.denominator, anotherTokenRate.numerator))
+		rightSide = new(big.Int).Mul(anotherTokenAmount, new(big.Int).Mul(anotherTokenRate.denominator, er.numerator))
 	}
 
-	return new(big.Int).Div(numerator, denominator), nil
+	return leftSide.Cmp(rightSide)
 }
 
-type CurrencyComparator struct {
-	exchangeRates    map[common.Address]*exchangeRate
-	_getExchangeRate func(*common.Address) (*exchangeRate, error)
+type CurrencyManager struct {
+	header *types.Header
+	state  vm.StateDB
+
+	exchangeRates    map[common.Address]*ExchangeRate                                        // map of exchange rates of the form (cGLD, token)
+	_getExchangeRate func(*common.Address, *types.Header, vm.StateDB) (*ExchangeRate, error) // function to obtain exchange rate from blockchain state
 }
 
-func NewComparator() *CurrencyComparator {
-	return newComparator(getExchangeRate)
+func NewManager(header *types.Header, state vm.StateDB) *CurrencyManager {
+	return newManager(GetExchangeRate, header, state)
 }
 
-func newComparator(_getExchangeRate func(*common.Address) (*exchangeRate, error)) *CurrencyComparator {
-	return &CurrencyComparator{
-		exchangeRates:    make(map[common.Address]*exchangeRate),
+func newManager(_getExchangeRate func(*common.Address, *types.Header, vm.StateDB) (*ExchangeRate, error), header *types.Header, state vm.StateDB) *CurrencyManager {
```

Comment:
this can be merged into `NewManager`

Reply:
not really, this is used internally for testing. test define a mock `_getExchangeRate` function. 

That's why this function exists, and it's private